opengever/maintenance/scripts/migrate_and_deactivate_meetings.py

```diff
@@ -0,0 +1,303 @@
+        allowed_states = ['dossier-state-resolved', 'dossier-state-inactive', 'dossier-state-active']
+        dossiers_in_bad_state = []
+        for brain in meeting_dossiers:
+            if brain.review_state not in allowed_states:
+                dossiers_in_bad_state.append(brain)
+
+        if dossiers_in_bad_state:
+            dossiers_in_bad_state_table = TextTable()
+            dossiers_in_bad_state_table.add_row(
+                (u"Path", u"Title", u"State"))
+
+            for brain in dossiers_in_bad_state:
+                dossiers_in_bad_state_table.add_row((
+                    brain.getPath(),
+                    brain.title.replace(",", ""),
+                    brain.review_state))
+
+            self.log_and_write_table(
+                dossiers_in_bad_state_table,
+                "Dossiers in bad review states",
+                "dossiers_in_bad_state")
+
+        if active_meetings.count() or active_proposals.count() or dossiers_in_bad_state:
+            raise PreconditionsError("Preconditions not satisfied")
+
+    def replace_meeting_dossier_with_normal_dossier(self):
+        meeting_dossiers = self.catalog.unrestrictedSearchResults(portal_type="opengever.meeting.meetingdossier")
+        to_delete = []
+        for brain in meeting_dossiers:
+            meeting_dossier = brain.getObject()
+
+            # create simple dossier
+            parent = aq_parent(meeting_dossier)
+            data = DexterityObjectDataExtractor(meeting_dossier).extract()
+            data = encode_after_json(data)
+            data[BASEDATA_KEY][u'portal_type'] = u'opengever.dossier.businesscasedossier'
+            del data[FIELDDATA_KEY][u'IMeetingDossier']
+            data[FIELDDATA_KEY][u'IBusinessCaseDossier'] = {}
+            data[FIELDDATA_KEY][u'IProtectDossier'] = {}
+            data = decode_for_json(data)
+            dossier = DexterityObjectCreator(data).create_in(parent)
+
+            # Move all content of meeting dossier to normal dossier
+            for obj in meeting_dossier.contentValues():
+                api.content.move(obj, dossier)
+
+            # update reference in meeting
+            meeting = meeting_dossier.get_meeting()
+            meeting.dossier_oguid = Oguid.for_object(dossier)
+
+            # update favorites
+            query = Favorite.query.by_object(meeting_dossier)
+            query.update({'oguid': Oguid.for_object(dossier)})
+
+            # Save state of the meeting dossier on the new dossier
+            migration_annotations = self.get_migration_annotations(dossier)
+            migration_annotations['state'] = api.content.get_state(meeting_dossier)
+
+            # We will delete the dossiers afterwards, as it otherwise seems
+            # to modify meeting_dossiers during iteration.
+            to_delete.append(meeting_dossier)
+
+        for obj in to_delete:
+            # delete meeting_dossier
+            api.content.delete(obj)
```

Comment:
I would probably do some high level logging here, e.g.
`Deleting meeting dossier {meeting_dossier_path} after moving contents to regular dossier {new_dossier_path}`

Reply:
Right thanks, I had actually planed to add more logging overall and seems I forgot. I'll use the progress logger everywhere and add some explicit logging what is deleted and where stuff gets moved to. I'll also add the path of the former meeting dossier into the annotations of the new dossier and check that the object being deleted is empty...